docs/Toy/toy/interpreter.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import annotations
+
+import operator
+from typing import Any
+from itertools import accumulate
+
+from dataclasses import dataclass
+
+from xdsl.dialects.builtin import TensorType, VectorType, ModuleOp
+from xdsl.interpreter import Interpreter, InterpreterFunctions, register_impls, impl
+from xdsl.utils.exceptions import InterpretationError
+
+from . import dialect as toy
+
+
+@dataclass
+class Tensor:
+    data: list[float]
+    shape: list[int]
+
+    def __format__(self, __format_spec: str) -> str:
+        prod_shapes: list[int] = list(
+            accumulate(reversed(self.shape), operator.mul))
+        assert prod_shapes[-1] == len(self.data)
+        result = '[' * len(self.shape)
+
+        for i, d in enumerate(self.data):
+            if i:
+                n = sum(not i % p for p in prod_shapes)
+                result += ']' * n
+                result += ', '
+                result += '[' * n
+            result += f'{d}'
+
+        result += ']' * len(self.shape)
+        return result
+
+
+@register_impls
+class ToyFunctions(InterpreterFunctions):
+
+    def run_toy_func(self, interpreter: Interpreter, name: str,
+                     args: tuple[Any, ...]) -> tuple[Any, ...]:
+        for op in interpreter.module.ops:
+            if isinstance(op, toy.FuncOp) and op.sym_name.data == name:
+                return self.run_func(interpreter, op, args)
+
+        raise InterpretationError(
+            f'Could not find toy function with name: {name}')
+
+    @impl(toy.PrintOp)
+    def run_print(self, interpreter: Interpreter, op: toy.PrintOp,
+                  args: tuple[Any, ...]) -> tuple[Any, ...]:
+        interpreter.print(f'{args[0]}')
+        return ()
+
+    @impl(toy.FuncOp)
+    def run_func(self, interpreter: Interpreter, op: toy.FuncOp,
```

Comment:
Is a toy func isolated from above? From how it seems here, it is not as it seems to behave more like a nested scope rather than a new scope. Is there enough functionality in the interpreter env to have several, unrelated environments?

Reply:
So you are saying that the structure of the pdl dialect does not allow this? Can't you just have environments that a `get_item` goes through and ones that it doesn't or something? I feel like we are implementing something here that will not be able to interpret any valid IR. If that is the case, we should state this somewhere very prominently.

Reply:
So you are saying that you assume that the given IR is in a valid form and, therefore, this cannot happen?
Feels like a precondition that should be mentioned (even though it is quite obvious).